[no-misused-promises] false positives with overloaded function

<!--
If you have a problem with a specific rule, please begin your issue title with [rulename] to make it easier to search for.
I.e. "[no-unused-vars] False positive when fooing the bar"

Please don't ignore this template.

If you ignore it, we're just going to respond asking you to fill it out, which wastes everyone's time.
The more relevant information you can include, the faster we can find the issue and fix it without asking you for more info.
-->

<!--
Are you opening an issue because the rule you're trying to use is not found?
🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨
1) Check the releases log: https://github.com/typescript-eslint/typescript-eslint/releases
    -  If the rule isn't listed there, then chances are it hasn't been released to the main npm tag yet.
2) Try installing the `canary` tag: `npm i @typescript-eslint/eslint-plugin@canary`.
    - The canary tag is built for every commit to main, so it contains the bleeding edge build.
3) If ESLint still can't find the rule, then consider reporting an issue.
-->

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

- https://github.com/AviVahl/typescript-eslint-bug
- default tsconfig from `tsc --init`
- minimal eslint config
- clone and `npm it` to see issue

**Expected Result**

mocha's `it` accepts a return promise in its callback type, so there should be no error.

**Actual Result**

```
typescript-eslint-bug/test/some-test.spec.ts
  2:32  error  Promise returned in function argument where a void return was expected  @typescript-eslint/no-misused-promises
```
mishandling of overloads defined on an interface?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.13.0` |
| `@typescript-eslint/parser`        | `5.13.0` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.10.0` |
| `node`                             | `16.14.0` |

--- a/packages/eslint-plugin/src/rules/no-misused-promises.ts
+++ b/packages/eslint-plugin/src/rules/no-misused-promises.ts
@@ -394,8 +394,12 @@ function voidFunctionParams(
   checker: ts.TypeChecker,
   node: ts.CallExpression | ts.NewExpression,
 ): Set<number> {
+  const thenableReturnIndices = new Set<number>();
   const voidReturnIndices = new Set<number>();
   const type = checker.getTypeAtLocation(node.expression);
+
+  // TODO(file bug on TypeScript): checker.getResolvedSignature prefers a () => void over a () => Promise<void>
+  // See https://github.com/typescript-eslint/typescript-eslint/issues/4609 (todo: comment in PR)
 
   for (const subType of tsutils.unionTypeParts(type)) {
     // Standard function calls and `new` have two different types of signatures
@@ -408,11 +412,20 @@ function voidFunctionParams(
           parameter,
           node.expression,
         );
-        if (isVoidReturningFunctionType(checker, node.expression, type)) {
+        if (isThenableReturningFunctionType(checker, node.expression, type)) {
+          thenableReturnIndices.add(index);
+        } else if (
+          !thenableReturnIndices.has(index) &&
+          isVoidReturningFunctionType(checker, node.expression, type)
+        ) {
           voidReturnIndices.add(index);
         }
       }
     }
+  }
+
+  for (const index of thenableReturnIndices) {
+    voidReturnIndices.delete(index);
   }
 
   return voidReturnIndices;
@@ -424,21 +437,42 @@ function isVoidReturningFunctionType(
   node: ts.Node,
   type: ts.Type,
 ): boolean {
-  let hasVoidReturningFunction = false;
-  let hasThenableReturningFunction = false;
   for (const subType of tsutils.unionTypeParts(type)) {
     for (const signature of subType.getCallSignatures()) {
       const returnType = signature.getReturnType();
+
+      // If a certain positional argument accepts both thenable and void returns,
+      // a promise-returning function is valid
+      if (tsutils.isThenableType(checker, node, returnType)) {
+        return false;
+      }
+
       if (tsutils.isTypeFlagSet(returnType, ts.TypeFlags.Void)) {
-        hasVoidReturningFunction = true;
-      } else if (tsutils.isThenableType(checker, node, returnType)) {
-        hasThenableReturningFunction = true;
+        return true;
       }
     }
   }
-  // If a certain positional argument accepts both thenable and void returns,
-  // a promise-returning function is valid
-  return hasVoidReturningFunction && !hasThenableReturningFunction;
+  return false;
+}
+
+/**
+ * @returns Whether type is a void-returning function.
+ */
+function isThenableReturningFunctionType(
+  checker: ts.TypeChecker,
+  node: ts.Node,
+  type: ts.Type,
+): boolean {
+  for (const subType of tsutils.unionTypeParts(type)) {
+    for (const signature of subType.getCallSignatures()) {
+      const returnType = signature.getReturnType();
+      if (tsutils.isThenableType(checker, node, returnType)) {
+        return true;
+      }
+    }
+  }
+
+  return false;
 }
 
 // Returns true if the expression is a function that returns a thenable

--- a/packages/eslint-plugin/src/rules/no-misused-promises.ts
+++ b/packages/eslint-plugin/src/rules/no-misused-promises.ts
@@ -431,7 +431,44 @@ function voidFunctionParams(
   return voidReturnIndices;
 }
 
-// Returns true if given type is a void-returning function.
+/**
+ * @returns Whether any call signature of the type has a thenable return type.
+ */
+function anySignatureIsThenableType(
+  checker: ts.TypeChecker,
+  node: ts.Node,
+  type: ts.Type,
+): boolean {
+  for (const signature of type.getCallSignatures()) {
+    const returnType = signature.getReturnType();
+    if (tsutils.isThenableType(checker, node, returnType)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/**
+ * @returns Whether type is a thenable-returning function.
+ */
+function isThenableReturningFunctionType(
+  checker: ts.TypeChecker,
+  node: ts.Node,
+  type: ts.Type,
+): boolean {
+  for (const subType of tsutils.unionTypeParts(type)) {
+    if (anySignatureIsThenableType(checker, node, subType)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/**
+ * @returns Whether type is a void-returning function.
+ */
 function isVoidReturningFunctionType(
   checker: ts.TypeChecker,
   node: ts.Node,
@@ -456,36 +493,13 @@ function isVoidReturningFunctionType(
 }
 
 /**
- * @returns Whether type is a void-returning function.
+ * @returns Whether expression is a function that returns a thenable.
  */
-function isThenableReturningFunctionType(
-  checker: ts.TypeChecker,
-  node: ts.Node,
-  type: ts.Type,
-): boolean {
-  for (const subType of tsutils.unionTypeParts(type)) {
-    for (const signature of subType.getCallSignatures()) {
-      const returnType = signature.getReturnType();
-      if (tsutils.isThenableType(checker, node, returnType)) {
-        return true;
-      }
-    }
-  }
-
-  return false;
-}
-
-// Returns true if the expression is a function that returns a thenable
 function returnsThenable(checker: ts.TypeChecker, node: ts.Node): boolean {
   const type = checker.getApparentType(checker.getTypeAtLocation(node));
 
-  for (const subType of tsutils.unionTypeParts(type)) {
-    for (const signature of subType.getCallSignatures()) {
-      const returnType = signature.getReturnType();
-      if (tsutils.isThenableType(checker, node, returnType)) {
-        return true;
-      }
-    }
+  if (anySignatureIsThenableType(checker, node, type)) {
+    return true;
   }
 
   return false;

--- a/packages/eslint-plugin/src/rules/no-misused-promises.ts
+++ b/packages/eslint-plugin/src/rules/no-misused-promises.ts
@@ -398,8 +398,8 @@ function voidFunctionParams(
   const voidReturnIndices = new Set<number>();
   const type = checker.getTypeAtLocation(node.expression);
 
-  // TODO(file bug on TypeScript): checker.getResolvedSignature prefers a () => void over a () => Promise<void>
-  // See https://github.com/typescript-eslint/typescript-eslint/issues/4609 (todo: comment in PR)
+  // We can't use checker.getResolvedSignature because it prefers an early '() => void' over a later '() => Promise<void>'
+  // See https://github.com/microsoft/TypeScript/issues/48077
 
   for (const subType of tsutils.unionTypeParts(type)) {
     // Standard function calls and `new` have two different types of signatures

--- a/packages/eslint-plugin/tests/rules/no-misused-promises.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-misused-promises.test.ts
@@ -276,6 +276,34 @@ declare const it: ItLike;
 it('', async () => {});
       `,
     },
+    {
+      code: `
+interface ItLike {
+  (name: string, callback: () => void): void;
+}
+interface ItLike {
+  (name: string, callback: () => Promise<void>): void;
+}
+
+declare const it: ItLike;
+
+it('', async () => {});
+      `,
+    },
+    {
+      code: `
+interface ItLike {
+  (name: string, callback: () => Promise<void>): void;
+}
+interface ItLike {
+  (name: string, callback: () => void): void;
+}
+
+declare const it: ItLike;
+
+it('', async () => {});
+      `,
+    },
   ],
 
   invalid: [
@@ -725,6 +753,46 @@ it('', async () => {});
       errors: [
         {
           line: 9,
+          messageId: 'voidReturnArgument',
+        },
+      ],
+    },
+    {
+      code: `
+interface ItLike {
+  (name: string, callback: () => number): void;
+}
+interface ItLike {
+  (name: string, callback: () => void): void;
+}
+
+declare const it: ItLike;
+
+it('', async () => {});
+      `,
+      errors: [
+        {
+          line: 11,
+          messageId: 'voidReturnArgument',
+        },
+      ],
+    },
+    {
+      code: `
+interface ItLike {
+  (name: string, callback: () => void): void;
+}
+interface ItLike {
+  (name: string, callback: () => number): void;
+}
+
+declare const it: ItLike;
+
+it('', async () => {});
+      `,
+      errors: [
+        {
+          line: 11,
           messageId: 'voidReturnArgument',
         },
       ],

--- a/packages/eslint-plugin/tests/rules/no-misused-promises.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-misused-promises.test.ts
@@ -252,6 +252,30 @@ const Component: any = () => null;
       `,
       filename: 'react.tsx',
     },
+    {
+      code: `
+interface ItLike {
+  (name: string, callback: () => Promise<void>): void;
+  (name: string, callback: () => void): void;
+}
+
+declare const it: ItLike;
+
+it('', async () => {});
+      `,
+    },
+    {
+      code: `
+interface ItLike {
+  (name: string, callback: () => void): void;
+  (name: string, callback: () => Promise<void>): void;
+}
+
+declare const it: ItLike;
+
+it('', async () => {});
+      `,
+    },
   ],
 
   invalid: [
@@ -684,6 +708,24 @@ const Component = (obj: O) => null;
         {
           line: 7,
           messageId: 'voidReturnAttribute',
+        },
+      ],
+    },
+    {
+      code: `
+interface ItLike {
+  (name: string, callback: () => number): void;
+  (name: string, callback: () => void): void;
+}
+
+declare const it: ItLike;
+
+it('', async () => {});
+      `,
+      errors: [
+        {
+          line: 9,
+          messageId: 'voidReturnArgument',
         },
       ],
     },

